source/MaterialXCore/Element.h

```diff
@@ -761,14 +761,16 @@ class MX_CORE_API Element : public std::enable_shared_from_this<Element>
     /// name, and attributes.
     string asString() const;
 
-    /// Resolve a reference to a named element at the root scope of this document,
-    /// taking the namespace at the scope of this element into account.
-    template <class T> shared_ptr<T> resolveRootNameReference(const string& name) const
+    /// Resolve a reference to a named element at the scope of the given parent,
+    /// taking the namespace at the scope of this element into account.  If no parent
+    /// is provided, then the root scope of the document is used.
+    template <class T> shared_ptr<T> resolveNameReference(const string& name, ConstElementPtr parent= nullptr) const
```

Comment:
Normally I would recommend that we add this new default argument to our Python binding for `resolveNameReference`, so that callers can use both the default and explicit forms of the method.

But I wonder if a better approach here would actually be to make `resolveNameReference` a protected method of Element, since it's not really a user-facing feature that would be expected as part of the Element API.  If we instead took this second approach, then I'd recommend that we simplify the changelist by removing the Python binding for this method as well.

What are your thoughts?

Reply:
Hi @jstone-lucasfilm 
The main implementation issue is that not all uses of this API are via derived classes so would need to work around this somehow. Perhaps for now, the Javascript and Python bindings can be removed first ?

Reply:
I've made it protected. ShaderRef upgrade used this code so I just copied a few lines over.